### PR TITLE
Automatically Generate Refill Date

### PR DESCRIPTION
**Pull request by GayathriVenkatesh**
# Description
Added feature to automatically fill in medication refill date, if not specified.

## Corresponding Issue
#1638 

# Screenshots

Screenshot from 2020-02-02 11-34-43

In this example, refill date is not filled in by the user. So it defaults to the second-last day of the dosage.

Screenshot from 2020-02-02 11-35-39

<br><br><br>
Screenshot from 2020-02-02 11-38-33

Here, the date is explicitly specified by the user, so it remains as it is.

Screenshot from 2020-02-02 11-39-03

---

Reviewing this pull request? Check out our [Code Review Practices](https://github.com/ifmeorg/ifme/wiki/Code-Review-Practices) guide if you haven't already!

### CODE REVIEW

**Comment by GayathriVenkatesh** [comment]
Hi @julianguyen, I worked on issue #1638 and have added screenshots for the same. However, there are a few codeclimate errors in the code. Is there a workaround to this? I tried a few things out but was unable to condense the code any further.

File: app/controllers/medications_controller.rb [hunk 1]
```diff
@@ -50,6 +50,9 @@ def edit
   def create
     @medication =
       Medication.new(medication_params.merge(user_id: current_user.id))
+
+    @medication.refill = @medication.refill.nil? ? Time.zone.now.strftime('%d/%m/%Y').to_date +
+      7 * ((@medication.total / @medication.dosage) / @medication.weekly_dosage.length) + ((@medication.total / @medication.dosage) % @medication.weekly_dosage.length) - 1 : @medication.refill
```
Comment (julianguyen) [inline app/controllers/medications_controller.rb]: I would would break apart this logic into a helper function. That should reduce the complexity of this create function and better separate the logic you're adding.

Comment (julianguyen) [reply app/controllers/medications_controller.rb]: I would use `if` and `unless` like Code Climate suggests instead of having multi-line ternary operators. That should also improve readability of the method.

You can run `rubocop` locally and it should throw these warnings.

Comment (GayathriVenkatesh) [reply app/controllers/medications_controller.rb]: Thanks! I have fixed those issues, now there is no error in codeclimate. However, the total-coverage and diff-coverage checks have failed. I did research a bit but I couldn't pinpoint the error. Would be great if you could look into it!

File: app/controllers/medications_controller.rb [hunk 2]
```diff
@@ -45,11 +45,24 @@ def edit
     redirect_to_medication(@medication)
   end
 
+  def add_refill(medicine)
```
Comment (julianguyen) [inline app/controllers/medications_controller.rb]: We should also write a unit test for this function!

**Comment by julianguyen** [comment]
Looks like we should have test coverage for this new feature! Let me know if you need any help writing tests :)

File: app/controllers/medications_controller.rb (hunk 2 above)
Comment (GayathriVenkatesh) [reply app/controllers/medications_controller.rb]: Hi @julianguyen, I am not quite familiar with writing tests for such a function. Would be great if someone could guide me through it!

Comment (julianguyen) [reply app/controllers/medications_controller.rb]: What I would do this update this part of the [test](https://github.com/ifmeorg/ifme/blob/master/spec/controllers/medications_controller_spec.rb#L126) to ensure that `@medication.refill` is the expected value.

**Comment by GayathriVenkatesh** [comment]
Hi @julianguyen, I updated the spec.rb file to ensure that the refill date is always before the last date of dosage. Could you please have a look and let me know if there's something I should change?

File: spec/controllers/medications_controller_spec.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
-
+Rails::logger.debug "Interesting stuff"
```
Comment (julianguyen) [inline spec/controllers/medications_controller_spec.rb]: We should remove this!

File: spec/controllers/medications_controller_spec.rb [hunk 4]
```diff
@@ -143,6 +143,9 @@ def post_create(medication_params)
         it 'has no validation errors' do
           post_create valid_medication_params
           expect(assigns(:medication).errors).to be_empty
+  
+          expect(assigns(:medication)[:refill]).to be_between("01-01-0001".to_date,
+            ((7 * assigns(:medication)[:total]) / (assigns(:medication)[:dosage] * assigns(:medication)[:weekly_dosage].count)).days.from_now)
```
Comment (julianguyen) [inline spec/controllers/medications_controller_spec.rb]: Is checking between 01-01-0001 and the number of days from when the days are calculated the best way of verifying this works? Why not set the created_at date to a value upon create and then check for the specific date in the test?

File: app/controllers/medications_controller.rb [hunk 5]
```diff
@@ -45,11 +45,24 @@ def edit
...
+    if medicine.refill.nil?
```
Comment (julianguyen) [inline app/controllers/medications_controller.rb]: We should also test when `medicine.refill` is not nill!

What also happens when we update a medication - can we change the refill values?

**Comment by julianguyen** [comment]
The Codeclimate coverage failures mean that we should add more tests! I mentioned above that we should test the nil case.

File: spec/controllers/medications_controller_spec.rb (hunk 4 above)
Comment (GayathriVenkatesh) [reply spec/controllers/medications_controller_spec.rb]: Screenshot from 2020-03-06 13-26-03

Hi @julianguyen, I tried changing this part so that it checks whether the refill date is between date of creation of medication and the maximum possible date (as calculated). However this throws the above error. Could you please guide me further?

File: app/controllers/medications_controller.rb (hunk 5 above)
Comment (GayathriVenkatesh) [reply app/controllers/medications_controller.rb]: Updating medication works fine, no issues on that front :)

File: spec/controllers/medications_controller_spec.rb (hunk 4 above)
Comment (julianguyen) [reply spec/controllers/medications_controller_spec.rb]: Ah so this error is happening because the medication's refill is being set to `01/01/2020` in `spec/factories/factories.rb`. We use this factory to generate a medication in this test.

I tested this feature locally and it doesn't work because the front-end prevents us from submitting the form if the refill field is left blank.

Comment (julianguyen) [reply spec/controllers/medications_controller_spec.rb]: To fix this, we'll just need to set this field as non-required (just in the front-end not back-end). I can push a commit to this PR to update this.

Comment (julianguyen) [reply spec/controllers/medications_controller_spec.rb]: As for getting the test to pass, we'll need to call `add_refill` when we call `Medication.new` instead of doing it after the new medication entry is created. I'll also update your PR for you.

File: app/controllers/medications_controller.rb (hunk 5 above)
Comment (julianguyen) [reply app/controllers/medications_controller.rb]: So another important question to consider is whether we want add_refill to be called when updating a medication.

Comment (julianguyen) [reply app/controllers/medications_controller.rb]: Actually, we'll have to call add_refill for updating. It's possible that the user would empty out the refill field in the form. I can update this too!

**Comment by julianguyen** [comment]
Can you get rid of the two recent commits you added - it undoes some changes I made that we need.

**Comment by julianguyen** [comment]
I am going to merge this if the tests pass and the security audit fails. Our master has the latest version of the vulnerable gems. Outside of this PR, you should rebase from our master in your forked master.
